src/WasmExecutor.cpp

```diff
@@ -0,0 +1,1583 @@
+#if WASM_DEBUG_LEVEL >= 2
+    const halide_dimension_t *dim = buf->dim;
+    const uint8_t *host = buf->host;
+
+    wdebug(0) << label << " = " << (void *) buf << " = {\n";
+    wdebug(0) << "  device = " << buf->device << "\n";
+    wdebug(0) << "  device_interface = " << buf->device_interface << "\n";
+    wdebug(0) << "  host = " << (void *) host << " = {\n";
+    if (host) {
+        wdebug(0) << "    " << (int) host[0] << ", " << (int) host[1] << ", " << (int) host[2] << ", " << (int) host[3] << "...\n" ;
+    }
+    wdebug(0) << "  }\n";
+    wdebug(0) << "  flags = " << buf->flags << "\n";
+    wdebug(0) << "  type = " << (int) buf->type.code << "," << (int) buf->type.bits << "," << buf->type.lanes << "\n";
+    wdebug(0) << "  dimensions = " << buf->dimensions << "\n";
+    wdebug(0) << "  dim = " << (void*) buf->dim << " = {\n";
+    for (int i = 0; i < buf->dimensions; i++) {
+        const auto &d = dim[i];
+        wdebug(0) << "    {" << d.min << "," << d.extent << "," << d.stride << "," << d.flags << "},\n";
+    }
+    wdebug(0) << "  }\n";
+    wdebug(0) << "  padding = " << buf->padding << "\n";
+    wdebug(0) << "}\n";
+#endif
+}
+
+void dump_wasmbuf(const Local<Context> &context, wasm32_ptr_t buf_ptr, const std::string &label) {
+#if WASM_DEBUG_LEVEL >= 2
+    internal_assert(buf_ptr);
+
+    uint8_t *base = get_wasm_memory_base(context);
+    wasm_halide_buffer_t *buf = (wasm_halide_buffer_t *) (base + buf_ptr);
+    halide_dimension_t *dim = buf->dim ? (halide_dimension_t *) (base + buf->dim) : nullptr;
+    uint8_t *host = buf->host ? (base + buf->host) : nullptr;
+
+    wdebug(0) << label << " = " << buf_ptr << " -> " << (void *) buf << " = {\n";
+    wdebug(0) << "  device = " << buf->device << "\n";
+    wdebug(0) << "  device_interface = " << buf->device_interface << "\n";
+    wdebug(0) << "  host = " << buf->host << " -> " << (void *) host << " = {\n";
+    if (host) {
+        wdebug(0) << "    " << (int) host[0] << ", " << (int) host[1] << ", " << (int) host[2] << ", " << (int) host[3] << "...\n" ;
+    }
+    wdebug(0) << "  }\n";
+    wdebug(0) << "  flags = " << buf->flags << "\n";
+    wdebug(0) << "  type = " << (int) buf->type.code << "," << (int) buf->type.bits << "," << buf->type.lanes << "\n";
+    wdebug(0) << "  dimensions = " << buf->dimensions << "\n";
+    wdebug(0) << "  dim = " << buf->dim << " -> " << (void*) dim << " = {\n";
+    for (int i = 0; i < buf->dimensions; i++) {
+        const auto &d = dim[i];
+        wdebug(0) << "    {" << d.min << "," << d.extent << "," << d.stride << "," << d.flags << "},\n";
+    }
+    wdebug(0) << "  }\n";
+    wdebug(0) << "  padding = " << buf->padding << "\n";
+    wdebug(0) << "}\n";
+#endif
+}
+
+static_assert(sizeof(halide_type_t) == 4, "halide_type_t");
+static_assert(sizeof(halide_dimension_t) == 16, "halide_dimension_t");
+static_assert(sizeof(wasm_halide_buffer_t) == 40, "wasm_halide_buffer_t");
+
+size_t size_in_bytes(const halide_buffer_t *b) {
```

Comment:
How does this differ from b->size_in_bytes()?

Reply:
...I forgot that we had some of these methods defined on halide_buffer_t (as well as Buffer<>), so, it doesn't. Removing.